File: app/src/androidTest/java/info/sanaebadi/tddandroidunittest/NoteDaoTest.java
package info.sanaebadi.tddandroidunittest;

import android.database.sqlite.SQLiteConstraintException;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;


import org.junit.Assert;
import org.junit.Rule;
import org.junit.Test;

import java.util.List;

import info.sanaebadi.tddandroidunittest.models.Note;
import info.sanaebadi.tddandroidunittest.util.LiveDataTestUtil;
import info.sanaebadi.tddandroidunittest.util.TestUtil;

import static org.junit.Assert.*;

public class NoteDaoTest extends NoteDatabaseTest {

    public static final String TEST_TITLE = "This is a test title";
    public static final String TEST_CONTENT = "This is some test content";
    public static final String TEST_TIMESTAMP = "08-2018";

    @Rule
    public InstantTaskExecutorRule rule = new InstantTaskExecutorRule();

    /*
        Insert, Read, Delete
     */
    @Test
    public void insertReadDelete() throws Exception{

        Note note = new Note(TestUtil.TEST_NOTE_1);

        // insert
        getNoteDao().insertNote(note).blockingGet(); // wait until inserted

        // read
        LiveDataTestUtil<List<Note>> liveDataTestUtil = new LiveDataTestUtil<>();
        List<Note> insertedNotes = liveDataTestUtil.getValue(getNoteDao().notes);

        assertNotNull(insertedNotes);

        assertEquals(note.content, insertedNotes.get(0).content);
        assertEquals(note.timestamp, insertedNotes.get(0).timestamp);
        assertEquals(note.title, insertedNotes.get(0).title);

        note.id = insertedNotes.get(0).id;
        assertEquals(note, insertedNotes.get(0));

        // delete
        getNoteDao().deleteNote(note).blockingGet();

        // confirm the database is empty
        insertedNotes = liveDataTestUtil.getValue(getNoteDao().notes);
        assertEquals(0, insertedNotes.size());
    }


    /*
        Insert, Read, Update, Read, Delete,
     */
    @Test
    public void insertReadUpdateReadDelete() throws Exception{

        Note note = new Note(TestUtil.TEST_NOTE_1);

        // insert
        getNoteDao().insertNote(note).blockingGet(); // wait until inserted

        // read
        LiveDataTestUtil<List<Note>> liveDataTestUtil = new LiveDataTestUtil<>();
        List<Note> insertedNotes = liveDataTestUtil.getValue(getNoteDao().notes);

        assertNotNull(insertedNotes);

        assertEquals(note.content, insertedNotes.get(0).content);
        assertEquals(note.timestamp, insertedNotes.get(0).timestamp);
        assertEquals(note.title, insertedNotes.get(0).title);

        note.id = insertedNotes.get(0).id;
        assertEquals(note, insertedNotes.get(0));

        // update
        note.title = TEST_TITLE;
        note.content = TEST_CONTENT;
        note.timestamp = TEST_TIMESTAMP;
        getNoteDao().updateNote(note).blockingGet();

        // read
        insertedNotes = liveDataTestUtil.getValue(getNoteDao().notes);

        assertEquals(TEST_TITLE, insertedNotes.get(0).title);
        assertEquals(TEST_CONTENT, insertedNotes.get(0).content);
        assertEquals(TEST_TIMESTAMP, insertedNotes.get(0).timestamp);

        note.id = insertedNotes.get(0).id;
        assertEquals(note, insertedNotes.get(0));

        // delete
        getNoteDao().deleteNote(note).blockingGet();

        // confirm the database is empty
        insertedNotes = liveDataTestUtil.getValue(getNoteDao().notes);
        assertEquals(0, insertedNotes.size());
    }



    /*
        Insert note with null title, throw exception
     */
    @Test(expected = SQLiteConstraintException.class)
    public void insert_nullTitle_throwSQLiteConstraintException() throws Exception{

        final Note note = new Note(TestUtil.TEST_NOTE_1);
        note.title = null;

        // insert
        getNoteDao().insertNote(note).blockingGet();
    }


    /*
        Insert, Update with null title, throw exception
     */

    @Test(expected = SQLiteConstraintException.class)
    public void updateNote_nullTitle_throwSQLiteConstraintException() throws Exception{

        Note note = new Note(TestUtil.TEST_NOTE_1);

        // insert
        getNoteDao().insertNote(note).blockingGet();

        // read
        LiveDataTestUtil<List<Note>> liveDataTestUtil = new LiveDataTestUtil<>();
        List<Note> insertedNotes = liveDataTestUtil.getValue(getNoteDao().notes);
        assertNotNull(insertedNotes);

        // update
        note = new Note(insertedNotes.get(0));
        note.title = null;
        getNoteDao().updateNote(note).blockingGet();

    }


}

File: app/src/androidTest/java/info/sanaebadi/tddandroidunittest/NoteDaoTest.kt
package info.sanaebadi.tddandroidunittest

import info.sanaebadi.tddandroidunittest.model.Note
import info.sanaebadi.tddandroidunittest.util.LiveDataTestUtil
import info.sanaebadi.tddandroidunittest.util.TestUtil
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Test


class NoteDaoTest : NoteDatabaseTest() {


    companion object {
        const val TITLE_TEST = "This is Title test"
        const val CONTENT_TEST = "This is Content test"
        const val TIMESTAMP_TEST = "04-2020"
    }


    // Insert , Read , Delete

    @Test
    @Throws(java.lang.Exception::class)
    internal fun insertReadDelete() {

        val note = Note(TestUtil.TEST_NOTE_1)

        //insert
        getNoteDao().insertNote(note)!!.blockingGet() //wait until inserted


        // read

        // read
        var liveDataTestUtil = LiveDataTestUtil<List<Note>>()
        var insertedNotes = liveDataTestUtil.getValue(getNoteDao().getNotes())!!

        assertNotNull(insertedNotes)
        assertEquals(note.content, insertedNotes[0].content)
        assertEquals(note.title, insertedNotes[0].title)
        assertEquals(note.timestamp, insertedNotes[0].timestamp)

        note.id = insertedNotes[0].id
        assertEquals(note, insertedNotes[0])


        //delete
        getNoteDao().deleteNote(note).blockingGet()

        //confirm the database is empty
        insertedNotes = liveDataTestUtil.getValue(getNoteDao().getNotes())!!
        assertEquals(0, insertedNotes.size)
    }


    //Insert , Read , Update , Delete

    @Test
    @Throws(java.lang.Exception::class)
    internal fun insertReadUpdateReadDelete() {

        val note = Note(TestUtil.TEST_NOTE_1)

        //insert
        getNoteDao().insertNote(note)!!.blockingGet() //wait until inserted


        // read

        // read
        var liveDataTestUtil = LiveDataTestUtil<List<Note>>()
        var insertedNotes = liveDataTestUtil.getValue(getNoteDao().getNotes())!!

        assertNotNull(insertedNotes)
        assertEquals(note.content, insertedNotes[0].content)
        assertEquals(note.title, insertedNotes[0].title)
        assertEquals(note.timestamp, insertedNotes[0].timestamp)

        note.id = insertedNotes[0].id
        assertEquals(note, insertedNotes[0])


        //delete
        getNoteDao().deleteNote(note).blockingGet()

        //update
        note.title = TITLE_TEST
        note.content = CONTENT_TEST
        note.timestamp = TIMESTAMP_TEST
        getNoteDao().updateNote(note).blockingGet()


        //read
        insertedNotes = liveDataTestUtil.getValue(getNoteDao().getNotes())

        assertEquals(TITLE_TEST, insertedNotes[0].title)
        assertEquals(CONTENT_TEST, insertedNotes[0].content)
        assertEquals(TIMESTAMP_TEST, insertedNotes[0].timestamp)


        note.id = insertedNotes[0].id
        assertEquals(note, insertedNotes[0])


        //confirm the database is empty
        insertedNotes = liveDataTestUtil.getValue(getNoteDao().getNotes())!!
        assertEquals(0, insertedNotes.size)
    }


    //Insert Note with null title throw exception


    //Insert , Update  Note with null title throw exception


}